{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "## Скачиваем собранный геном бактерии (все скаффолды)"
      ],
      "metadata": {
        "id": "vZ-g_7FcfMwD"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Hqj0t9L_ea65",
        "outputId": "a1f90822-d59c-4b9b-df95-19c1412f6b5e"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2022-10-17 18:21:20--  https://raw.githubusercontent.com/sonishko/hse22_hw2/main/data/scaffolds.fa\n",
            "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 3955483 (3.8M) [text/plain]\n",
            "Saving to: ‘scaffolds.fa’\n",
            "\n",
            "scaffolds.fa        100%[===================>]   3.77M  --.-KB/s    in 0.07s   \n",
            "\n",
            "2022-10-17 18:21:20 (52.3 MB/s) - ‘scaffolds.fa’ saved [3955483/3955483]\n",
            "\n"
          ]
        }
      ],
      "source": [
        "!wget https://raw.githubusercontent.com/sonishko/hse22_hw2/main/data/scaffolds.fa\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Необходимо скачать GeneMarkS-2 и активационный ключ для него http://opal.biology.gatech.edu/GeneMark/license_download.cgi"
      ],
      "metadata": {
        "id": "1Kzn2lmBkAuR"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget http://topaz.gatech.edu/GeneMark/tmp/GMtool_18P7P/gms2_linux_64.tar.gz\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "yPQpoqDigrpj",
        "outputId": "d8b55bf2-1eea-49df-ad4c-2e688f79bb8a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2022-10-17 18:21:32--  http://topaz.gatech.edu/GeneMark/tmp/GMtool_18P7P/gms2_linux_64.tar.gz\n",
            "Resolving topaz.gatech.edu (topaz.gatech.edu)... 143.215.161.100\n",
            "Connecting to topaz.gatech.edu (topaz.gatech.edu)|143.215.161.100|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 11261753 (11M) [application/x-gzip]\n",
            "Saving to: ‘gms2_linux_64.tar.gz’\n",
            "\n",
            "gms2_linux_64.tar.g 100%[===================>]  10.74M  5.87MB/s    in 1.8s    \n",
            "\n",
            "2022-10-17 18:21:34 (5.87 MB/s) - ‘gms2_linux_64.tar.gz’ saved [11261753/11261753]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget http://topaz.gatech.edu/GeneMark/tmp/GMtool_18P7P/gm_key_64.gz"
      ],
      "metadata": {
        "id": "CJFD4Pr-hA3Z",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "83cf7085-e6eb-4f6c-bf54-11492ba039eb"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2022-10-17 18:21:39--  http://topaz.gatech.edu/GeneMark/tmp/GMtool_18P7P/gm_key_64.gz\n",
            "Resolving topaz.gatech.edu (topaz.gatech.edu)... 143.215.161.100\n",
            "Connecting to topaz.gatech.edu (topaz.gatech.edu)|143.215.161.100|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 87 [application/x-gzip]\n",
            "Saving to: ‘gm_key_64.gz’\n",
            "\n",
            "\rgm_key_64.gz          0%[                    ]       0  --.-KB/s               \rgm_key_64.gz        100%[===================>]      87  --.-KB/s    in 0s      \n",
            "\n",
            "2022-10-17 18:21:39 (11.4 MB/s) - ‘gm_key_64.gz’ saved [87/87]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!gzip -d gm_key_64.gz\n",
        "!tar -xzvf gms2_linux_64.tar.gz\n",
        "!cp  -v  gm_key_64   ~/.gmhmmp2_key\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "IWSwU6nKhWpd",
        "outputId": "6d9b5846-110c-45ae-c375-21a0106ef8ed"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "gms2_linux_64/\n",
            "gms2_linux_64/gms2.pl\n",
            "gms2_linux_64/README_GMS2.txt\n",
            "gms2_linux_64/mgm_4.mod\n",
            "gms2_linux_64/gmhmmp2\n",
            "gms2_linux_64/INSTALL\n",
            "gms2_linux_64/mgm_11.mod\n",
            "gms2_linux_64/biogem\n",
            "gms2_linux_64/LICENSE\n",
            "gms2_linux_64/mgm_25.mod\n",
            "gms2_linux_64/compp\n",
            "gms2_linux_64/mgm_15.mod\n",
            "'gm_key_64' -> '/root/.gmhmmp2_key'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Установка E-utilities (для скачивания последовательностей из NCBI)\n"
      ],
      "metadata": {
        "id": "1ezHknjWkKbq"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!sh -c \"$(curl -fsSL ftp://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/install-edirect.sh)\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OCjg19tmksVA",
        "outputId": "2ce95e34-8d74-4ae8-9dc7-84dfb14852ac"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Entrez Direct has been successfully downloaded and installed.\n",
            "\n",
            "In order to complete the configuration process, please execute the following:\n",
            "\n",
            "  echo \"export PATH=\\${PATH}:/root/edirect\" >> ${HOME}/.bashrc\n",
            "\n",
            "or manually edit the PATH variable assignment in your .bashrc file.\n",
            "\n",
            "Would you like to do that automatically now? [y/N]\n",
            "y\n",
            "OK, done.\n",
            "\n",
            "To activate EDirect for this terminal session, please execute the following:\n",
            "\n",
            "export PATH=${PATH}:${HOME}/edirect\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Скачиваем данные по близкородственной бактерии T.oleivorans"
      ],
      "metadata": {
        "id": "9JdvjcDVlBzm"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Геном, последовательности генов (нт) и белков (протеом) для бактерии Thalassolituus oleivorans MIL-1\n",
        "https://www.ncbi.nlm.nih.gov/nuccore/HF680312\n"
      ],
      "metadata": {
        "id": "XKq7OKHrlC2A"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!$HOME/edirect/efetch -db nuccore -id HF680312 -format gb  >  T_oleivorans_MIL_1.gbk\n",
        "!$HOME/edirect/efetch -db nuccore -id HF680312 -format gene_fasta  >  T_oleivorans_MIL_1.genes.fasta\n",
        "!$HOME/edirect/efetch -db nuccore -id HF680312 -format fasta_cds_aa  >  T_oleivorans_MIL_1.proteins.fasta"
      ],
      "metadata": {
        "id": "fwAeQ5OflLda"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Установка BLAST"
      ],
      "metadata": {
        "id": "G9-gFtjFlzJm"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get update\n",
        "!apt-get install ncbi-blast+"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KTE5SKKUl1oD",
        "outputId": "dcf87ff3-73aa-4083-addd-34780a6da332"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\r0% [Working]\r            \rGet:1 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease [3,626 B]\n",
            "\r0% [Connecting to archive.ubuntu.com (185.125.190.39)] [Waiting for headers] [1\r0% [Connecting to archive.ubuntu.com (185.125.190.39)] [Waiting for headers] [C\r0% [1 InRelease gpgv 3,626 B] [Connecting to archive.ubuntu.com (185.125.190.39\r                                                                               \rGet:2 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]\n",
            "Hit:3 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic InRelease\n",
            "Hit:4 http://archive.ubuntu.com/ubuntu bionic InRelease\n",
            "Get:5 http://archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]\n",
            "Hit:6 http://ppa.launchpad.net/cran/libgit2/ubuntu bionic InRelease\n",
            "Get:7 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bionic InRelease [15.9 kB]\n",
            "Get:8 http://archive.ubuntu.com/ubuntu bionic-backports InRelease [83.3 kB]\n",
            "Hit:9 http://ppa.launchpad.net/graphics-drivers/ppa/ubuntu bionic InRelease\n",
            "Ign:10 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease\n",
            "Hit:11 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease\n",
            "Hit:12 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release\n",
            "Get:13 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 Packages [3,454 kB]\n",
            "Get:14 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bionic/main amd64 Packages [44.0 kB]\n",
            "Get:15 http://archive.ubuntu.com/ubuntu bionic-updates/restricted amd64 Packages [1,232 kB]\n",
            "Fetched 5,009 kB in 3s (1,702 kB/s)\n",
            "Reading package lists... Done\n",
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "The following package was automatically installed and is no longer required:\n",
            "  libnvidia-common-460\n",
            "Use 'apt autoremove' to remove it.\n",
            "The following additional packages will be installed:\n",
            "  ncbi-data\n",
            "The following NEW packages will be installed:\n",
            "  ncbi-blast+ ncbi-data\n",
            "0 upgraded, 2 newly installed, 0 to remove and 22 not upgraded.\n",
            "Need to get 13.1 MB of archives.\n",
            "After this operation, 66.7 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 ncbi-data all 6.1.20170106-2 [3,645 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu bionic/universe amd64 ncbi-blast+ amd64 2.6.0-1 [9,446 kB]\n",
            "Fetched 13.1 MB in 1s (9,914 kB/s)\n",
            "Selecting previously unselected package ncbi-data.\n",
            "(Reading database ... 123941 files and directories currently installed.)\n",
            "Preparing to unpack .../ncbi-data_6.1.20170106-2_all.deb ...\n",
            "Unpacking ncbi-data (6.1.20170106-2) ...\n",
            "Selecting previously unselected package ncbi-blast+.\n",
            "Preparing to unpack .../ncbi-blast+_2.6.0-1_amd64.deb ...\n",
            "Unpacking ncbi-blast+ (2.6.0-1) ...\n",
            "Setting up ncbi-data (6.1.20170106-2) ...\n",
            "Setting up ncbi-blast+ (2.6.0-1) ...\n",
            "Processing triggers for hicolor-icon-theme (0.17-2) ...\n",
            "Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Скачиваем БД белков SwissProt"
      ],
      "metadata": {
        "id": "gqkE0SSrmZB1"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KI-nG_zKmeUb",
        "outputId": "1e7e9667-2037-4037-bcde-39ee99573370"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2022-10-17 18:24:47--  https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz\n",
            "Resolving ftp.uniprot.org (ftp.uniprot.org)... 128.175.240.195\n",
            "Connecting to ftp.uniprot.org (ftp.uniprot.org)|128.175.240.195|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 91303010 (87M) [application/x-gzip]\n",
            "Saving to: ‘uniprot_sprot.fasta.gz’\n",
            "\n",
            "uniprot_sprot.fasta 100%[===================>]  87.07M   110MB/s    in 0.8s    \n",
            "\n",
            "2022-10-17 18:24:48 (110 MB/s) - ‘uniprot_sprot.fasta.gz’ saved [91303010/91303010]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!gzip -d uniprot_sprot.fasta.gz"
      ],
      "metadata": {
        "id": "2md5ESsVmhS-"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Устанавливаем пакет seqkit"
      ],
      "metadata": {
        "id": "e4O7GJhYmrOw"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://github.com/shenwei356/seqkit/releases/download/v2.1.0/seqkit_linux_amd64.tar.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "eOx-RTYJmuKN",
        "outputId": "bb078d1a-5676-4b97-8523-6347e2c3f640"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2022-10-17 18:24:56--  https://github.com/shenwei356/seqkit/releases/download/v2.1.0/seqkit_linux_amd64.tar.gz\n",
            "Resolving github.com (github.com)... 140.82.112.3\n",
            "Connecting to github.com (github.com)|140.82.112.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/52715040/ea4018ec-70e8-40ac-a4aa-3326b04e07c4?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20221017%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20221017T182456Z&X-Amz-Expires=300&X-Amz-Signature=868661430c3c8642a02dcdaaba0b89dbb919e92ad11de88e1bf6942c94d2f7c0&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=52715040&response-content-disposition=attachment%3B%20filename%3Dseqkit_linux_amd64.tar.gz&response-content-type=application%2Foctet-stream [following]\n",
            "--2022-10-17 18:24:56--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/52715040/ea4018ec-70e8-40ac-a4aa-3326b04e07c4?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20221017%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20221017T182456Z&X-Amz-Expires=300&X-Amz-Signature=868661430c3c8642a02dcdaaba0b89dbb919e92ad11de88e1bf6942c94d2f7c0&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=52715040&response-content-disposition=attachment%3B%20filename%3Dseqkit_linux_amd64.tar.gz&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.111.133, 185.199.108.133, 185.199.109.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.111.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 6650850 (6.3M) [application/octet-stream]\n",
            "Saving to: ‘seqkit_linux_amd64.tar.gz’\n",
            "\n",
            "seqkit_linux_amd64. 100%[===================>]   6.34M  31.2MB/s    in 0.2s    \n",
            "\n",
            "2022-10-17 18:24:57 (31.2 MB/s) - ‘seqkit_linux_amd64.tar.gz’ saved [6650850/6650850]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!tar -xzvf seqkit_linux_amd64.tar.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Yghfyqkum4aF",
        "outputId": "1a6e67a9-d888-4771-9f65-5686ce02355a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "seqkit\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!chmod a+x seqkit"
      ],
      "metadata": {
        "id": "Qc5m5OmEm7E_"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Выполнение задания"
      ],
      "metadata": {
        "id": "JYLmia-9naqB"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Предсказать расположение белок-кодирующих генов в геноме нашей бактерии"
      ],
      "metadata": {
        "id": "ad1RRVeUneZE"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "* Предсказать белок-кодирующие гены в собранном геноме нашей бактерии с помощью ab initio программы GeneMarkS – комманда gms2.pl\n",
        "* По умолчанию координаты предсказанных генов сохраняются в файле gms2.lst\n",
        "* Обратите внимание, что GeneMarkS может выдавать последовательности генов (опция --fnn), а также автоматически их транслировать в белки (--faa)"
      ],
      "metadata": {
        "id": "iKrtO7Eknhb0"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!./gms2_linux_64/gms2.pl   --seq scaffolds.fa  --genome-type bacteria  --fnn genes.fasta  --faa proteins.fasta"
      ],
      "metadata": {
        "id": "HZnlRS3-nlIh"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Определяем функции белков нашей бактерии через сравнение с белками из T.oleivorans MIL-1\n",
        "\n",
        "* Сравнить все белки (протеом) нашей бактерии с аннотированными белками T. oleivorans:\n",
        "* Протеом T. oleivorans MIL-1 находится в файле T_oleivorans_MIL_1.proteins.fasta\n",
        "* Индексируем белки T.oleivorans MIL-1:"
      ],
      "metadata": {
        "id": "iK9nEYDtNHUV"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!makeblastdb  -dbtype prot  -in T_oleivorans_MIL_1.proteins.fasta  -out T_oleivorans_MIL_1.proteins"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "K8wJrWTZNNUt",
        "outputId": "f5514449-b6dc-4ce8-ff39-d1ee2261e21f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "\n",
            "Building a new DB, current time: 10/17/2022 18:29:40\n",
            "New DB name:   /content/T_oleivorans_MIL_1.proteins\n",
            "New DB title:  T_oleivorans_MIL_1.proteins.fasta\n",
            "Sequence type: Protein\n",
            "Keep MBits: T\n",
            "Maximum file size: 1000000000B\n",
            "Adding sequences from FASTA; added 3666 sequences in 0.106353 seconds.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "* Определеям какие белки из нашей бактерии похожи на белки из T_oleivorans:\n"
      ],
      "metadata": {
        "id": "8Eu5oYz4NXzx"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!blastp  -query proteins.fasta  -db T_oleivorans_MIL_1.proteins  -evalue 1e-10  -outfmt 6  >  scaffolds.hits_from_MIL_1.txt"
      ],
      "metadata": {
        "id": "RemKq6u1Nb_u"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Определяем функции белков нашей бактерии через поиск по БД SwissProt\n",
        "\n",
        "* Смотрим, какие белки нашей бактерии (из Баренцева моря) не имели похожих белков из бактерии T.oleivorans MIL-1\n",
        "* Для этого получаем список уникальных номеров генов бактерии из Барнцева моря, коротые имели хотя бы один похожий белок бактерии из Сицилли\n"
      ],
      "metadata": {
        "id": "WxL2XhumNhbZ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!cut -f 1 scaffolds.hits_from_MIL_1.txt | sort -n | uniq > proteins.with_hits_from_MIL_1.txt"
      ],
      "metadata": {
        "id": "GBejxJixNpJL"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!./seqkit grep  --invert-match  -f proteins.with_hits_from_MIL_1.txt  proteins.fasta   -o proteins.without_MIL_1.fasta"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "U-M_tBtBNzl7",
        "outputId": "60b29d14-c0db-4e58-85f9-fc927f292ec6"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[INFO]\u001b[0m 3324 patterns loaded from file\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "* Проверяем, что кол-ва белков сходятся\n",
        "* Общее кол-во белков в нашей бактерии:\n"
      ],
      "metadata": {
        "id": "aB02WiGgN3UO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!grep '>' proteins.fasta | wc -l"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "uPa6OKlUN6Xu",
        "outputId": "6872225a-c94c-40fd-a11d-7b49a7065eaa"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "3602\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "* Кол-во белков, которые имели схожие белки из бактерии MIL-1"
      ],
      "metadata": {
        "id": "cc1rWHhsN89N"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wc -l proteins.with_hits_from_MIL_1.txt"
      ],
      "metadata": {
        "id": "Bn9HNuAaOAdW"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!grep '>' proteins.without_MIL_1.fasta | wc -l"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "l86-ccm6OC8z",
        "outputId": "711d3204-8ab0-4783-e08b-026990755ecb"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "278\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "* Сравнить все белки (протеом) нашей бактерии с аннотированными белками из БД SwissProt\n",
        "* Индексируем белки SwissProt:"
      ],
      "metadata": {
        "id": "u--isFLtObwU"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!makeblastdb  -dbtype prot  -in uniprot_sprot.fasta  -out uniprot_sprot"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-PTLODWNOc87",
        "outputId": "d42f1df1-7238-4656-b7a1-1489093ede2c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "\n",
            "Building a new DB, current time: 10/17/2022 18:34:42\n",
            "New DB name:   /content/uniprot_sprot\n",
            "New DB title:  uniprot_sprot.fasta\n",
            "Sequence type: Protein\n",
            "Keep MBits: T\n",
            "Maximum file size: 1000000000B\n",
            "Adding sequences from FASTA; added 568363 sequences in 13.3782 seconds.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "* Определеям какие белки из нашей бактерии похожи на белки из БД SwissProt (программа работает около 10 мин):"
      ],
      "metadata": {
        "id": "C9N3_2LaOgjW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!time  blastp  -query proteins.without_MIL_1.fasta  -db uniprot_sprot  -evalue 1e-10  -outfmt 6  >  scaffolds.hits_from_SwissProt.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KckihgwGOk-b",
        "outputId": "7c925249-81e0-4169-b35f-f699429cfd45"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "real\t9m29.355s\n",
            "user\t9m27.930s\n",
            "sys\t0m0.401s\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "* Посмотрим, сколько белков удалось аннотировать через поиск по SwissProt:"
      ],
      "metadata": {
        "id": "mGX20dYwQ92A"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!cut -f 1 scaffolds.hits_from_SwissProt.txt | sort | uniq | wc -l"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "AM8d0pCQQ3Yu",
        "outputId": "6bf22d0e-4233-4562-c2c7-18899eea0e9a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "53\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Загружаем файлы, которые нам потребуются для аннотации генома, на github\n",
        "\n",
        "* gms2.lst  --  координаты всех предсказанных генов\n",
        "* proteins.fasta  --  аминокислотные последовательности всех предсказанных генов\n",
        "* scaffolds.hits_from_MIL_1.txt  --  информация о схожести белков из нашей бактерии с белками из бактерии MIL-1\n",
        "* scaffolds.hits_from_SwissProt.txt  -- информация о схожести белков из нашей бактерии с белками из БД SwissProt"
      ],
      "metadata": {
        "id": "2hNS1rKKRD8o"
      }
    }
  ]
}